File: apps/scala/src/main/scala/machinelearning/chapter11/CrossValidation.scala
package machinelearning.chapter11

import machinelearning.chapter02.{Features, Preprocessing}

/** 交差検証の 1 回分の分け方。行の位置（0 始まり）を訓練データとテストデータに分けて持つ。
  *
  * @param train
  *   訓練データの行の位置
  * @param test
  *   テストデータの行の位置
  */
case class Fold(train: Vector[Int], test: Vector[Int])

/** K 分割交差検証。 */
object CrossValidation:

  /** シード付きの乱数で行を並べ替え、nSplits 個のテストデータに分ける。
    *
    * 件数が割り切れないときは、余りを先頭の分割から 1 件ずつ配る。第 2 章の `Preprocessing.shuffle` を使うので、Java 版と同じ並びになる。
    */
  def kFold(nSamples: Int, nSplits: Int, seed: Long): Vector[Fold] =
    require(nSplits >= 1, "分割の数は 1 以上にしてください")
    require(nSamples >= nSplits, "件数は分割の数以上でなければなりません")
    val positions = Preprocessing.shuffle((0 until nSamples).toVector, seed)
    val sizes =
      (0 until nSplits).toVector.map(i =>
        nSamples / nSplits + (if i < nSamples % nSplits then 1 else 0)
      )
    sizes
      .scanLeft(0)(_ + _)
      .lazyZip(sizes)
      .map { (from, size) =>
        val test = positions.slice(from, from + size)
        Fold(positions.filterNot(test.toSet), test)
      }
      .toVector

  /** 分割ごとに新しいモデルを作って訓練データで学習し、テストデータの予測を評価関数で採点する。
    *
    * スコアは遅延評価の LazyList で返す。取り出した分だけ学習する。
    */
  def crossValidate[T](
      makeModel: () => Model[T],
      x: Vector[Features],
      t: Vector[T],
      folds: Vector[Fold],
      metric: Metric[T]
  ): LazyList[Double] =
    folds.to(LazyList).map { fold =>
      val fitted = makeModel().fit(pick(x, fold.train), pick(t, fold.train))
      metric(pick(t, fold.test), fitted.predict(pick(x, fold.test)))
    }

  private def pick[A](values: Vector[A], positions: Vector[Int]): Vector[A] =
    positions.map(values)

File: apps/go/internal/chapter11/crossvalidation.go
package chapter11

import (
	"fmt"
	"slices"

	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter02"
)

// Fold は 1 回分の分け方。訓練データとテストデータの行番号を持つ。
// 行そのものではなく行番号を持つので、特徴量と正解ラベルのどちらにも同じ分け方を使える。
type Fold struct {
	Train []int
	Test  []int
}

// KFold は行数を nSplits 個に分け、順番にテストデータにした分け方を返す。
// 余りは先の分け方に 1 件ずつ配るので、分け方の大きさの差は 1 件までになる。
func KFold(nSamples, nSplits int, seed int64) ([]Fold, error) {
	if nSplits < 2 {
		return nil, fmt.Errorf("分割数は 2 以上にしてください: %d", nSplits)
	}

	if nSamples < nSplits {
		return nil, fmt.Errorf("行数 %d が分割数 %d より少ないです", nSamples, nSplits)
	}

	positions := make([]int, nSamples)
	for i := range positions {
		positions[i] = i
	}

	shuffled := chapter02.Shuffle(positions, seed)
	folds := make([]Fold, 0, nSplits)
	start := 0

	for i := range nSplits {
		size := nSamples / nSplits
		if i < nSamples%nSplits {
			size++
		}

		test := slices.Clone(shuffled[start : start+size])
		train := make([]int, 0, nSamples-size)

		train = append(train, shuffled[:start]...)
		train = append(train, shuffled[start+size:]...)

		folds = append(folds, Fold{Train: train, Test: test})
		start += size
	}

	return folds, nil
}

// pick は行番号の順に値を取り出す。
func pick[E any](values []E, positions []int) ([]E, error) {
	picked := make([]E, 0, len(positions))

	for _, position := range positions {
		if position < 0 || position >= len(values) {
			return nil, fmt.Errorf("行番号が範囲の外です: %d", position)
		}

		picked = append(picked, values[position])
	}

	return picked, nil
}

// CrossValidate は分け方ごとに新しいモデルを学習し、評価関数の値を並べて返す。
// モデルは makeModel で毎回作り直す。前の分け方の学習が残ったモデルを使い回さないようにするため。
func CrossValidate[T any](
	makeModel func() (Model[T], error),
	x []chapter02.Features,
	t []T,
	folds []Fold,
	metric Metric[T],
) ([]float64, error) {
	if err := requireSameSize(x, t); err != nil {
		return nil, err
	}

	scores := make([]float64, 0, len(folds))

	for _, fold := range folds {
		score, err := validateOne(makeModel, x, t, fold, metric)
		if err != nil {
			return nil, err
		}

		scores = append(scores, score)
	}

	return scores, nil
}

// validateOne は 1 つの分け方で学習して評価する。
func validateOne[T any](
	makeModel func() (Model[T], error),
	x []chapter02.Features,
	t []T,
	fold Fold,
	metric Metric[T],
) (float64, error) {
	model, err := makeModel()
	if err != nil {
		return 0, err
	}

	xTrain, err := pick(x, fold.Train)
	if err != nil {
		return 0, err
	}

	tTrain, err := pick(t, fold.Train)
	if err != nil {
		return 0, err
	}

	if err := model.Fit(xTrain, tTrain); err != nil {
		return 0, err
	}

	xTest, err := pick(x, fold.Test)
	if err != nil {
		return 0, err
	}

	tTest, err := pick(t, fold.Test)
	if err != nil {
		return 0, err
	}

	predicted, err := model.Predict(xTest)
	if err != nil {
		return 0, err
	}

	return metric(tTest, predicted)
}

// Mean は平均。交差検証の分け方ごとのスコアをまとめるのに使う。
func Mean(values []float64) (float64, error) {
	if len(values) == 0 {
		return 0, fmt.Errorf("値が 1 件もありません")
	}

	sum := 0.0
	for _, value := range values {
		sum += value
	}

	return sum / float64(len(values)), nil
}
